src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+
+      chunk_row_idx = 0;
+      std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+    }
+  }
+
+  if (chunk_row_idx != 0) {
+    emit_chunk();
+  }
+
+  return out_table;
+}
+
+std::shared_ptr<const Table> SetUnion::_analyze_input() {
+  Assert(_input_table_left()->column_count() == _input_table_right()->column_count(),
+         "Input tables must have the same layout. Column count mismatch.");
+
+  // Later code relies on input tables containing columns
+  if (_input_table_left()->column_count() == 0) {
+    return _input_table_left();
+  }
+
+  /**
+   * Check the column layout (column names and column types)
+   */
+  for (ColumnID::base_type column_idx = 0; column_idx < _input_table_left()->column_count(); ++column_idx) {
+    Assert(_input_table_left()->column_type(ColumnID{column_idx}) ==
+               _input_table_right()->column_type(ColumnID{column_idx}),
+           "Input tables must have the same layout. Column type mismatch.");
+    Assert(_input_table_left()->column_name(ColumnID{column_idx}) ==
+               _input_table_right()->column_name(ColumnID{column_idx}),
+           "Input tables must have the same layout. Column name mismatch.");
+  }
+
+  /**
+   * Later code relies on both tables having > 0 rows. If one doesn't, we can just return the other as the result of
+   * the operator
+   */
+  if (_input_table_left()->row_count() == 0) {
+    return _input_table_right();
+  }
+  if (_input_table_right()->row_count() == 0) {
+    return _input_table_left();
+  }
+
+  /**
+   * Both tables must contain only ReferenceColumns
+   */
+  Assert(_input_table_left()->get_type() == TableType::References &&
+             _input_table_right()->get_type() == TableType::References,
+         "SetUnion doesn't support non-reference tables yet");
+
+  /**
+   * Identify the column segments (verification that this is the same for all chunks happens in the #if IS_DEBUG block
+   * below)
+   */
+  const auto add_column_segments = [&](const auto& table) {
+    auto current_pos_list = std::shared_ptr<const PosList>();
+    const auto& first_chunk = table->get_chunk(ChunkID{0});
+    for (auto column_id = ColumnID{0}; column_id < table->column_count(); ++column_id) {
+      const auto column = first_chunk.get_column(column_id);
+      const auto ref_column = std::dynamic_pointer_cast<const ReferenceColumn>(column);
+      auto pos_list = ref_column->pos_list();
+
+      if (current_pos_list != pos_list) {
+        current_pos_list = pos_list;
+        _column_segment_begins.emplace_back(column_id);
+      }
+    }
+  };
+
+  add_column_segments(_input_table_left());
+  add_column_segments(_input_table_right());
+
+  std::sort(_column_segment_begins.begin(), _column_segment_begins.end());
+  const auto unique_end_iter = std::unique(_column_segment_begins.begin(), _column_segment_begins.end());
+  _column_segment_begins.resize(std::distance(_column_segment_begins.begin(), unique_end_iter));
```

Comment:
I think this could use a comment on what ` _column_segment_begins` actually holds. Especially why there are entries for both left and right tables in the same list and why they are unique. The comment in the header does not address this. Maybe an example helps?

Reply:
added an example.